File: package/umt_python/src/math/solve_equation.py
from .addition import addition
from .division import division
from .multiplication import multiplication
from .round_of import round_of
from .subtraction import subtraction


def solve_equation(
    coefficients: list[list[float]],
    constants: list[float],
) -> list[float]:
    """
    Solves a system of linear equations using Gaussian elimination.

    Args:
        coefficients: Matrix of coefficients.
        constants: Vector of constants.

    Returns:
        Solution vector.

    Example:
        >>> solve_equation([[1, 1], [1, 2]], [4, 10])
        [-2.0, 6.0]
    """
    n = len(constants)

    # Make copies to avoid modifying input
    coef = [row[:] for row in coefficients]
    const = constants[:]

    for index in range(n):
        # Find the max element in the column (partial pivoting)
        max_element = abs(coef[index][index])
        max_row = index
        for row in range(index + 1, n):
            if abs(coef[row][index]) > max_element:
                max_element = abs(coef[row][index])
                max_row = row

        # Swap the row with the max element to the top of the matrix
        for col in range(index, n):
            coef[max_row][col], coef[index][col] = coef[index][col], coef[max_row][col]

        const[max_row], const[index] = const[index], const[max_row]

        # Perform elimination on the rows below
        for row in range(index + 1, n):
            factor = division(coef[row][index], coef[index][index])

            const[row] = subtraction(const[row], multiplication(factor, const[index]))

            for col in range(index, n):
                coef[row][col] = subtraction(
                    coef[row][col], multiplication(factor, coef[index][col])
                )

    solution: list[float] = []

    # Back substitute to find the solution
    for row in range(n - 1, -1, -1):
        total = 0.0
        for col in range(row + 1, n):
            total = addition(
                total, multiplication(coef[row][col], solution[n - col - 1])
            )

        solution.append(division(subtraction(const[row], total), coef[row][row]))

    # Return the solution vector with values rounded to 1 decimal place
    return [round_of(value, 1) for value in reversed(solution)]

File: package/umt_rust/src/math/solve_equation.rs
use super::umt_round_of;

/// Solves a system of linear equations using Gaussian elimination.
///
/// Uses Gaussian elimination with partial pivoting to solve a system of linear equations.
/// The solution is rounded to 1 decimal place.
///
/// # Arguments
///
/// * `coefficients` - Matrix of coefficients.
/// * `constants` - Vector of constants.
///
/// # Returns
///
/// Solution vector.
///
/// # Examples
///
/// ```
/// use umt_rust::math::umt_solve_equation;
///
/// // Solves the system:
/// // x + y = 4
/// // x + 2y = 10
/// let result = umt_solve_equation(
///     &mut vec![vec![1.0, 1.0], vec![1.0, 2.0]],
///     &mut vec![4.0, 10.0]
/// );
/// assert_eq!(result, vec![-2.0, 6.0]);
/// ```
#[allow(clippy::ptr_arg, clippy::needless_range_loop)]
pub fn umt_solve_equation(coefficients: &mut Vec<Vec<f64>>, constants: &mut Vec<f64>) -> Vec<f64> {
    let n = constants.len();

    for index in 0..n {
        // Find the max element in the column (partial pivoting)
        let mut max_element = coefficients[index][index].abs();
        let mut max_row = index;

        for row in (index + 1)..n {
            if coefficients[row][index].abs() > max_element {
                max_element = coefficients[row][index].abs();
                max_row = row;
            }
        }

        // Swap the row with the max element to the top of the matrix
        for col in index..n {
            let temp = coefficients[max_row][col];
            coefficients[max_row][col] = coefficients[index][col];
            coefficients[index][col] = temp;
        }

        constants.swap(max_row, index);

        // Perform elimination on the rows below
        for row in (index + 1)..n {
            let factor = coefficients[row][index] / coefficients[index][index];

            constants[row] -= factor * constants[index];

            for col in index..n {
                coefficients[row][col] -= factor * coefficients[index][col];
            }
        }
    }

    let mut solution: Vec<f64> = vec![0.0; n];

    // Back substitute to find the solution
    for row in (0..n).rev() {
        let mut sum = 0.0;
        for col in (row + 1)..n {
            sum += coefficients[row][col] * solution[col];
        }

        solution[row] = (constants[row] - sum) / coefficients[row][row];
    }

    // Return the solution vector with values rounded to 1 decimal place
    solution.iter().map(|&v| umt_round_of(v, 1)).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solve_equation_basic() {
        // x + y = 4
        // x + 2y = 10
        let result = umt_solve_equation(
            &mut vec![vec![1.0, 1.0], vec![1.0, 2.0]],
            &mut vec![4.0, 10.0],
        );
        assert_eq!(result, vec![-2.0, 6.0]);
    }

    #[test]
    fn test_solve_equation_3x3() {
        // 2x + y + z = 9
        // x + 2y + z = 8
        // x + y + 2z = 7
        let result = umt_solve_equation(
            &mut vec![
                vec![2.0, 1.0, 1.0],
                vec![1.0, 2.0, 1.0],
                vec![1.0, 1.0, 2.0],
            ],
            &mut vec![9.0, 8.0, 7.0],
        );
        assert_eq!(result, vec![3.0, 2.0, 1.0]);
    }

    #[test]
    fn test_solve_equation_single() {
        // 2x = 6
        let result = umt_solve_equation(&mut vec![vec![2.0]], &mut vec![6.0]);
        assert_eq!(result, vec![3.0]);
    }
}
